central/administration/events/datastore/internal/writer/writer_impl.go

```diff
@@ -0,0 +1,142 @@
+package writer
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/administration/events/datastore/internal/store"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/administration/events"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+var (
+	_ Writer = (*writerImpl)(nil)
+
+	log           = logging.LoggerForModule()
+	maxWriterSize = 1000
+	eventSAC      = sac.ForResource(resources.Administration)
+)
+
+type writerImpl struct {
+	mutex sync.Mutex
+
+	buffer map[string]*storage.AdministrationEvent
+	store  store.Store
+}
+
+func (c *writerImpl) readNoLock(id string) (*storage.AdministrationEvent, bool) {
+	event, found := c.buffer[id]
+	return event, found
+}
+
+func (c *writerImpl) writeNoLock(event *storage.AdministrationEvent) {
+	c.buffer[event.GetId()] = event
+}
+
+func (c *writerImpl) resetNoLock() {
+	c.buffer = make(map[string]*storage.AdministrationEvent)
+}
+
+func (c *writerImpl) Upsert(ctx context.Context, event *events.AdministrationEvent) error {
+	if event == nil {
+		return errox.InvalidArgs.CausedBy("empty event")
+	}
+
+	if err := sac.VerifyAuthzOK(eventSAC.WriteAllowed(ctx)); err != nil {
+		return err
+	}
+
+	var isBufferFull bool
+	concurrency.WithLock(&c.mutex, func() {
+		isBufferFull = len(c.buffer) >= maxWriterSize
+	})
+
+	// If the buffer is full, first flush and clear the buffer. This ensures concurrent callers won't receive
+	// unexpected errors and additional logic on their side to flush and retry.
+	if isBufferFull {
```

Comment:
This can lead to the TOCTOU race (Time-of-Check Time-of-Use) and multiple flushes from concurrent goroutines when only one was necessary. Not a big deal here but still not a reason to have this bad pattern in the code. I'd suggest to pull flushing into `flushNoLock()` and call it under critical section together with the check.

To be pedantic, this alone does not entirely remove TOCTOU here since the buffer can fill in-between checking/flushing and writing. So either we need to leave a comment here or below that the buffer size can be overflown and it's alright, or do all operations under a single critical section which would be my preference (note that event enrichment does not have to be under critical section.